src/server/services/procedures/salio/salio.js

```diff
@@ -0,0 +1,749 @@
+     * @param {bool=} state Initial check state of the checkbox (defaults to false).
+     * @param {String} id Name of the checkbox.
+     * @param {String} event Name of the event to fire when the button is pressed.
+     * @param {string} text The text to display next to the checkbox
+     * @returns {boolean} True if the action is successful, false otherwise.
+     */
+    SalIO.prototype.addCheckbox = function (sensor, password, x, y, checkColor=this.getColor(66, 135, 245), textColor=this.getColor(0, 0, 0), state=false, id, event, text) {
+        arguments[4] = checkColor;
+        arguments[5] = textColor;
+        return this._passToSensor('addCheckbox', arguments);
+    };
+    /**
+     * Add a custom toggle switch to the device, functionally equivalent to a checkbox but styled as a switch.
+     * @param {string} sensor name of the sensor (matches at the end)
+     * @param {string} password current password for the sensor
+     * @param {BoundedNumber<0, 100>} x X position of the top left corner of the button (percentage).
+     * @param {BoundedNumber<0, 100>} y Y position of the top left corner of the button (percentage).
+     * @param {Number=} checkColor Color code of the check box itself (defaults to light blue).
+     * @param {Number=} textColor Color code of the text (if any) (defaults to black).
+     * @param {bool=} state Initial check state of the toggle switch (defaults to false).
+     * @param {String} id name of the toggle switch
+     * @param {String} event name of the event to raise for click events
+     * @param {string} text The text to display next to the toggle switch
+     * @returns {boolean} True if the action is successful, false otherwise.
+     */
+    SalIO.prototype.addToggleswitch = function (sensor, password, x, y, checkColor=this.getColor(66, 135, 245), textColor=this.getColor(0, 0, 0), state=false, id, event, text) {
+        arguments[4] = checkColor;
+        arguments[5] = textColor;
+        return this._passToSensor('addToggleswitch', arguments);
+    };
+    /**
+     * Add a custom radio button to the device.
+     * A radio button is like a checkbox, but they are arranged into groups.
+     * Only one radio button in each group may be checked.
+     * @param {string} sensor name of the sensor (matches at the end)
+     * @param {string} password current password for the sensor
+     * @param {BoundedNumber<0, 100>} x X position of the top left corner of the button (percentage).
+     * @param {BoundedNumber<0, 100>} y Y position of the top left corner of the button (percentage).
+     * @param {Number=} checkColor Color code of the check box itself (defaults to light blue).
+     * @param {Number=} textColor Color code of the text (if any) (defaults to black).
+     * @param {bool=} state Initial check state of the radio button (defaults to false) - All false in a group is OK, but you should avoid setting multiple explicitly to true.
+     * @param {String} id name of the radio button
```

Comment:
Maybe we should remove this argument and just return the ID. I am not sure if we should throw an error or return false on error... Thoughts?

Reply:
Well for that at least, having the user specify it gives options for meaningful names like `startButton`. If server generated the names they'd also need to use a bunch of variables to hold the results. Also I was imagining multiple people connecting to the same phone and using the same gui elements (e.g. for a group chat function), in which case we'd also need to transmit the auto-generated names to the other projects. I figured it'd just be easiest if everyone attempted to make `sharedText` (first will succeed) and all refer to it with the same name that they know up front.

Reply:
Gotcha. I hadn't been imagining multiple users programming the same phone; it certainly could be cumbersome to share the IDs this way. One thing that I don't like about passing it as an argument is that there are a lot of arguments here which can be a little overwhelming so it is nice to try to remove arguments that could be inferred, created, etc. It's probably fine for now.

What about error handling? I think throwing a meaningful error message might be better than simply returning true/false (though I know that approach is used in RoboScape). It would be nice to be able to differentiate between failed RPCs because the phone with that ID is not found or some other failure occurred.